[SPARK-49011][SQL] Improve exceptions thrown from parser/interpreter

dusantism-db opened the pull request:
### What changes were proposed in this pull request?
Introduced a new class SqlScriptingException, which is thrown during SQL script parsing/interpreting, and contains information about the line number on which the error occured.


### Why are the changes needed?
Users should know which line of their script caused an error.


### Does this PR introduce _any_ user-facing change?
Users will now see a line number on some of their error messages.


### How was this patch tested?
No new tests required, existing tests are updated.


### Was this patch authored or co-authored using generative AI tooling?
No

dusantism-db committed:
add SqlScriptingException and update sql scripting errors functions

dusantism-db committed:
add SqlScriptingException and update sql scripting errors functions

dusantism-db committed:
update tests and fix exception formatting

dusantism-db committed:
update tests and fix exception formatting

dusantism-db committed:
fix lint

dusantism-db committed:
fix lint

wayneguow committed:
[MINOR][DOCS] Fix typos in `docs/sql-migration-guide.md`

### What changes were proposed in this pull request?

This PR aims to fix typos in `docs/sql-migration-guide.md`.

### Why are the changes needed?

Fix typos.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Pass GA.

### Was this patch authored or co-authored using generative AI tooling?

No.

Closes #47530 from wayneguow/sql_typos.

Authored-by: Wei Guo <guow93@gmail.com>
Signed-off-by: Hyukjin Kwon <gurwls223@apache.org>

zhengruifeng committed:
[SPARK-49053][PYTHON][ML] Make model save/load helper functions accept spark session

### What changes were proposed in this pull request?
Make model save/load helper functions accept spark session

### Why are the changes needed?
1, avoid unnecessary spark session creations;
2, to be consistent with scala side changes: https://github.com/apache/spark/pull/47467 and https://github.com/apache/spark/pull/47477

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
CI

### Was this patch authored or co-authored using generative AI tooling?
No

Closes #47527 from zhengruifeng/py_ml_save_metadata_spark.

Authored-by: Ruifeng Zheng <ruifengz@apache.org>
Signed-off-by: Hyukjin Kwon <gurwls223@apache.org>

gengliangwang committed:
[SPARK-49054][SQL] Column default value should support current_* functions

### What changes were proposed in this pull request?

This is a regression between Spark 3.5.0 and Spark 4.
The following queries work on Spark 3.5.0 while fails on latest master branch:
```
CREATE TABLE test_current_user(i int, s string) USING parquet;
ALTER TABLE test_current_user ALTER COLUMN s SET DEFAULT current_user()
```
```
CREATE TABLE test_current_user(i int, s string default current_user()) USING parquet
INSERT INTO test_current_user (i) VALUES ((0));
```

This PR is to complete fixing this by eagerly executing finish-analysis and constant-folding rules before checking whether the expression is foldable and resolved.
### Why are the changes needed?

Bug fix

### Does this PR introduce _any_ user-facing change?

No
### How was this patch tested?

New UTs
### Was this patch authored or co-authored using generative AI tooling?

No

Closes #47529 from gengliangwang/finishAnalysis.

Authored-by: Gengliang Wang <gengliang@apache.org>
Signed-off-by: Gengliang Wang <gengliang@apache.org>

zedtang committed:
[SPARK-48762][SQL] Introduce clusterBy DataFrameWriter API for Python

### What changes were proposed in this pull request?

Introduce clusterBy DataFrameWriter API for Python.

Also fix the issue that `listColumns` doesn't support `V1Table`.

### Why are the changes needed?

Introduce more ways for users to interact with clustered tables.

### Does this PR introduce _any_ user-facing change?

Yes, it introduces a new PySpark DataFrame API to specify clustering columns during write operations.

### How was this patch tested?

New unit tests.

### Was this patch authored or co-authored using generative AI tooling?

No.

Closes #47452 from zedtang/clusterby-python-api.

Authored-by: Jiaheng Tang <jiaheng.tang@databricks.com>
Signed-off-by: Hyukjin Kwon <gurwls223@apache.org>

davidm-db commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/parser/AstBuilder.scala:
```diff
@@ -181,10 +183,15 @@ class AstBuilder extends DataTypeAstBuilder with SQLConfHelper
-          bl.multipartIdentifier().getText, el.multipartIdentifier().getText)
+        withOrigin(bl) {
+          throw SqlScriptingException.labelsMismatch(
+          CurrentOrigin.get, bl.multipartIdentifier().getText, el.multipartIdentifier().getText)
```
nit: this row should be indented for one tab.

davidm-db commented on sql/catalyst/src/main/scala/org/apache/spark/sql/errors/SqlScriptingException.scala:
```diff
@@ -17,40 +17,77 @@
+    cause = cause,
+    messageParameters = messageParameters
+  ) {
+
```
you can omit {}, it's not needed. also, throughout the code we don't leave `)` in the separate row, so it's better to do `messageParameters = messageParameters)` here

davidm-db commented on sql/catalyst/src/main/scala/org/apache/spark/sql/errors/SqlScriptingException.scala:
```diff
@@ -17,40 +17,77 @@
+    origin: Origin,
+    varName: String,
+    lineNumber: String
+  ): Throwable = {
```
same as in the comment above, usually we leave this part of the code in the line above: `lineNumber: String): Throwable = {`

davidm-db commented on sql/catalyst/src/main/scala/org/apache/spark/sql/errors/SqlScriptingException.scala:
```diff
@@ -17,40 +17,77 @@
-  def variableDeclarationOnlyAtBeginning(varName: String, lineNumber: String): Throwable = {
-    new SparkException(
+  def variableDeclarationOnlyAtBeginning(
+    origin: Origin,
```
we usually indent parameters for 4 spaces, and lines in the function for 2 spaces, so when you apply all the comments, this function should look like
```
def variableDeclarationOnlyAtBeginning(
    origin: Origin,
    varName: String,
    lineNumber: String): Throwable = {
  new SqlScriptingException(...)
}
```

davidm-db commented on sql/catalyst/src/main/scala/org/apache/spark/sql/errors/SqlScriptingException.scala:
```diff
@@ -17,40 +17,77 @@
+    errorClass: String,
+    messageParameters: Map[String, String]
+  ): String = {
+    val prefix = if (origin.line.isEmpty) "" else s"[LINE:${origin.line.get}] "
```
with `Options`, this line can be simplified and written in a more "functional" manner, something like:
`val lineNumberPrefix = origin.line.map(l => s"[LINE:${l}] ").getOrElse("")`

davidm-db commented on sql/catalyst/src/main/scala/org/apache/spark/sql/errors/SqlScriptingException.scala:
```diff
@@ -17,40 +17,77 @@
  * of the SQL Scripting Language interpreter.
  */
-private[sql] object SqlScriptingErrors extends QueryErrorsBase {
+private[sql] object SqlScriptingException {
```
I'm thinking maybe we should leave `SqlScriptingErrors` instead of using `SqlScriptingException` object - if you look at the `SparkException` as an example, they use object only for some internal stuff and exceptions, parameter formatting, etc. while various error classes are used to construct concrete `SparkExceptions`.

In such case, I would move `SqlScriptingException` to something like `org.apache.spark.sql.exceptions` - I know it doesn't exist at the moment, but I think we should create the package since I don't see any more meaningful place to put the new exception type.

davidm-db commented on sql/catalyst/src/main/scala/org/apache/spark/sql/errors/SqlScriptingException.scala:
```diff
@@ -17,40 +17,77 @@
+    errorClass: String,
+    cause: Throwable,
+    messageParameters: Map[String, String] = Map.empty)
+  extends SparkException(
```
This is fine, but I think it can be further simplified, we can extend `Exception(message, cause)` the same way that `SparkException` does it, and implementing a few simple but required override functions. reasoning: there are properties in `SparkException` that we actually don't need.

vicennial committed:
[SPARK-49059][CONNECT] Move `SessionHolder.forTesting(...)` to the test package

### What changes were proposed in this pull request?

Moves `SessionHolder.forTetsing(...)` into the package; the new and equivalent method lies in `SparkConnectTestUtils.createDummySessionHolder(...)`.

### Why are the changes needed?

The `SessionHolder.forTesting(...)` method is widely used in several Spark Connect testing suites. However, this test-specific code is located in the source code, and incorrect use may lead to the creation of a partial/buggy Spark Connect session during production runtime.

### Does this PR introduce _any_ user-facing change?

Yes.
Code running in the Server JVM will no longer be able to create a dummy Session Holder during live execution by calling `SessionHolder.forTesting(...)`.
Note that usage of this during live execution is unwanted behaviours as it would lead to creation of a partial `SessionHolder` instance.

### How was this patch tested?

Existing tests.

### Was this patch authored or co-authored using generative AI tooling?

No

Closes #47536 from vicennial/SPARK-49059.

Authored-by: vicennial <venkata.gudesa@databricks.com>
Signed-off-by: Hyukjin Kwon <gurwls223@apache.org>

ericm-db committed:
[SPARK-49031] Implement validation for the TransformWithStateExec operator using OperatorStateMetadataV2

### What changes were proposed in this pull request?

Implementing validation for the TransformWithStateExec operator, so that it can't restart with a different TimeMode and OutputMode, or invalid State Variable transformations.

### Why are the changes needed?

If there is an invalid change to the query after a restart, we want the query to fail.

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Unit tests

### Was this patch authored or co-authored using generative AI tooling?

No

Closes #47508 from ericm-db/validation.

Authored-by: Eric Marnadi <eric.marnadi@databricks.com>
Signed-off-by: Jungtaek Lim <kabhwan.opensource@gmail.com>

ilicmarkodb committed:
[SPARK-49003][SQL] Fix interpreted code path hashing to be collation aware

### What changes were proposed in this pull request?
Changed hash function to be collation aware. This change is just for interpreted code path. Codegen hashing was already collation aware.

### Why are the changes needed?
We were getting the wrong hash for collated strings because our hash function only used the binary representation of the string. It didn't take collation into account.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Added tests to `HashExpressionsSuite.scala`

### Was this patch authored or co-authored using generative AI tooling?
No

Closes #47502 from ilicmarkodb/ilicmarkodb/fix_string_hash.

Lead-authored-by: Marko <marko.ilic@databricks.com>
Co-authored-by: Marko Ilic <marko.ilic@databricks.com>
Co-authored-by: Marko Ilić <marko.ilic@databricks.com>
Signed-off-by: Wenchen Fan <wenchen@databricks.com>

uros-db committed:
[SPARK-48725][SQL] Integrate CollationAwareUTF8String.lowerCaseCodePoints into string expressions

### What changes were proposed in this pull request?
Use `CollationAwareUTF8String.lowerCaseCodePoints` logic to properly lowercase strings according to UTF8_LCASE collation, instead of relying on `UTF8String.toLowerCase()` method calls.

### Why are the changes needed?
Avoid correctness issues with respect to code-point logic in UTF8_LCASE (arising when Java to performs string lowercasing) and ensure consistent results.

### Does this PR introduce _any_ user-facing change?
Yes, collation aware string function implementations will now rely on `CollationAwareUTF8String` string lowercasing for UTF8_LCASE collation, instead of `UTF8String` logic (which resorts to Java's implementation).

### How was this patch tested?
Existing tests, with some new cases in `CollationSupportSuite`.
6 expressions are affected by this change: `contains`, `instr`, `find_in_set`, `replace`, `locate`, `substr_index`.

### Was this patch authored or co-authored using generative AI tooling?
No.

Closes #47132 from uros-db/lcase-cp.

Authored-by: Uros Bojanic <157381213+uros-db@users.noreply.github.com>
Signed-off-by: Wenchen Fan <wenchen@databricks.com>

itholic committed:
[SPARK-49056][SQL] ErrorClassesJsonReader cannot handle null properly

### What changes were proposed in this pull request?

This PR proposes to make `ErrorClassesJsonReader` handle null properly

### Why are the changes needed?

When `ErrorClassesJsonReader` takes null for `getErrorMessage` method, it cannot handle null properly so raises `INTERNAL_ERROR`.

For example, given error class example below:

```json
{
  "MISSING_PARAMETER" : {
    "message" : [
      "Parameter ${param} is missing."
    ]
  }
}
```

and run:

```scala
getErrorMessage("MISSING_PARAMETER", Map("param" -> null))
```

**Before**
```
[INTERNAL_ERROR] Undefined error message parameter for error class: 'MISSING_PARAMETER', MessageTemplate: Parameter ${param} is missing., Parameters: Map(param -> null) SQLSTATE: XX000
org.apache.spark.SparkException: [INTERNAL_ERROR] Undefined error message parameter for error class: 'MISSING_PARAMETER', MessageTemplate: Parameter ${param} is missing., Parameters: Map(param -> null) SQLSTATE: XX000
```

**After**
```
[MISSING_PARAMETER] Parameter null is missing.
```

### Does this PR introduce _any_ user-facing change?

No API changes, but this PR improves the user-facing error message as described above.

### How was this patch tested?

Added UT.

### Was this patch authored or co-authored using generative AI tooling?

No.

Closes #47531 from itholic/SPARK-49056.

Authored-by: Haejoon Lee <haejoon.lee@databricks.com>
Signed-off-by: Haejoon Lee <haejoon.lee@databricks.com>

wForget committed:
[SPARK-49067][SQL] Move utf-8 literal into internal methods of UrlCodec class

### What changes were proposed in this pull request?

Move utf-8 literals in url encode/decode functions to internal methods of UrlCodec class

### Why are the changes needed?

Remove unnecessary constant expressions

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

existing unit tests

### Was this patch authored or co-authored using generative AI tooling?

No

Closes #47544 from wForget/url_decode.

Authored-by: wforget <643348094@qq.com>
Signed-off-by: Kent Yao <yao@apache.org>

panbingkun committed:
[SPARK-49064][BUILD] Upgrade Kafka to 3.8.0

### What changes were proposed in this pull request?
The pr aims to upgrade `kafka` from `3.7.1` to `3.8.0`.

### Why are the changes needed?
https://downloads.apache.org/kafka/3.8.0/RELEASE_NOTES.html

### Does this PR introduce _any_ user-facing change?
No.

### How was this patch tested?
Pass GA.

### Was this patch authored or co-authored using generative AI tooling?
No.

Closes #47540 from panbingkun/SPARK-49064.

Authored-by: panbingkun <panbingkun@baidu.com>
Signed-off-by: Dongjoon Hyun <dhyun@apple.com>

dusantism-db committed:
fix formatting

dusantism-db committed:
sqlscriptingexception extends exception, not sparkexception

dusantism-db committed:
create exceptions package

dusantism-db committed:
Merge branch 'dusantism-db/sql-scripting-exception' of github.com:dusantism-db/spark into sql-scripting-exception

davidm-db commented:
@dusantism-db Let's close this PR so people don't get confused with the proper one by accident.
